spec/controllers/vm_infra_controller_spec.rb

```diff
@@ -328,6 +328,36 @@
     expect(response.status).to eq(200)
   end
 
+  context 'transform VM dialog' do
+    let(:dialog)          { FactoryGirl.create(:dialog) }
+    let(:resource_action) { FactoryGirl.create(:resource_action, :dialog => dialog) }
+
+    before do
+      allow(Dialog).to receive(:find_by).and_return(dialog)
+      allow(dialog).to receive(:resource_actions).and_return([resource_action])
+      allow(dialog).to receive(:buttons).and_return('')
+    end
```

Comment:
@smelamud you should not mock anything in the models at all:
* The `find_by` should work without mocking and if not, you're probably missing something from the controller's context 
* The `resource_actions` should work out of the box
* I don't really understand why the `buttons` is necessary here, in production it probably returns `nil` or an `array` and `FactoryGirl` ensures this feature in testing too.

Reply:
@skateman I checked line-by-line what works and what doesn't. These are necessary. I can send you exceptions if you want. For example, without `buttons` I get:
```
Error cau
ght: [ActionView::Template::Error] undefined method `split' for nil:NilClass
/home/smelamud/manageiq/plugins/manageiq-ui-classic/app/views/layouts/_x_dialog
_buttons.html.haml:6:in `_plugins_manageiq_ui_classic_app_views_layouts__x_dial
og_buttons_html_haml__4322103688760694089_47447439105460'
```
Sure, I don't have experience with this. So, if you see something that I missed in the controller or somewhere else, please, explain.

Reply:
@smelamud yup, the problem with the `buttons` here is that a dialog can't really exist without a button and the view where the error appears is designed with this constrain and your test fails. You will probably need to create some button factories and assign them into the dialog. I will dig into the models and paste the solution here soon.

Reply:
You're trying to mock [this line](https://github.com/ManageIQ/manageiq-ui-classic/blob/master/app/controllers/vm_infra_controller.rb#L101) with `find_by`. You can pass the same `:label` to the factory. The `dialogs` is a string field in the DB, it can be passed to the factory. Also you're using the `let` without referring to the variable in the spec, so the mocks aren't actually created. In such cases you should go with `let!` that forces the evaluation of the block afterwards.

```
    let(:dialog)           { FactoryGirl.create(:dialog, :buttons => 'submit', :label => 'Transform VM') }
    let!(:resource_action) { FactoryGirl.create(:resource_action, :dialog => dialog) }
```

With this code you don't need any mocks...

Reply:
@skateman Thanks!